### ISSUE TITLE
atomic problem in sequential scan

### ISSUE THREAD

**Issue by yingjunwu** [cmu-db/peloton#106]
Currently, we cannot guarantee atomicity when performing sequential scan. Let us now consider the scenario that a transaction `txn_A` updates a tuple with key=10. This action essentially contains three sub-actions: (1) create a local version `V_new` with the latest value assigned; (2) make the local version `V_new` globally visible by setting its begin and end timestamp; (3) change the older version `V_old`'s end timestamp. If a transaction `txn_B` performs sequential scan to search for tuples with key=10, it may find nothing! The key reason is that `txn_A`'s sub-actions (2) and (3) are not performed atomically. It is possible that `txn_B` first reads `V_new` and finds it invisible, and then `txn_A` performs (2) and (3), and after that `txn_B` reads `V_old` and finds it also invisible! This problem requires special support from garbage-collection component.

**Comment by jarulraj** [comment]
How can the garbage-collection component help ?

**Comment by yingjunwu** [comment]
@jarulraj sure it will work. I am coordinating with GC team to address this problem.

**Comment by jarulraj** [comment]
@yingjunwu Can you please fix this issue in your GC PR ?

**Comment by yingjunwu** [comment]
This is similar to phantom problem and let's hard to fix it in the GC PR. I will try to fix it in the next PR.

**Comment by jarulraj** [comment]
@yingjunwu How to ensure atomicity of sub-actions (2) and (3)?

**Comment by yingjunwu** [comment]
@jarulraj One approach is to scan the index nodes instead of tile groups. But I think this approach can cause higher cache miss rate, and @pmenon may be unhappy with that.

In fact, we suffer from two different levels of phantom problems:

1. Tuple-level phantom problem, which occurs when a transaction is scanning a table while another one is inserting/deleting a tuple from the table?
2. Version-level phantom problem, which occurs when a transaction is scanning a table while another one is inserting/deleting a version from the table?

We may need to take a closer look at this problem.

**Comment by yingjunwu** [comment]
The atomic problem mentioned in this issue directly results in the bug reported in #1154, but in different forms. Let's consider the following SQL statements:

```
CREATE TABLE foo(id integer PRIMARY KEY, year integer);
INSERT INTO foo VALUES(1, 100);
UPDATE foo SET id = 1000 WHERE year = 100; 
```

The `update` statement does the following things: 
- invoke the `seq_scan_evaluator` (as there's no index on `year` attribute) to find all the matching tuples in each tile group of the table `foo`;
- remove all the matching tuples from the table by creating a new empty version (which is tagged as deleted);
- create a new tuple with id set to 1000.

This statement (which updates a tuple's primary-key attributes) is more complex than statements which update tuples' non-primary-key attributes in that it needs to not only add a new empty version for the updated tuple but also create a new tuple with the new primary key.

As Peloton executes queries tile-group-by-tile-group, it is possible that the newly created tuple is allocated onto a tile group that is not yet processed by the `update` statement. If this happens, Peloton continues to scan those unprocessed tile groups, find the newly created tuple, and delete it, causing the bug reported in #1154.

**Comment by poojanilangekar** [comment]
@yingjunwu The assertion didn't fail when it was trying to update the same transaction. It is finding the tuple from the previous transaction. If it was the case that the transaction was seeing its own version of the tuple, it wouldn't fail on line 641 of `timestamp_ordering_transaction_context.cpp` because the `last_reader_commit_id` would be the same.  

The bigger question I have is, why are we updating the index even before the transaction commits? Shouldn't the

**Comment by poojanilangekar** [comment]
Fixed in #1272 & #1263
